Wasapi passthrough rework

kevmitch opened the pull request:
Make passthrough more user-friendly and correct.

wm4 commented:
Lots of changes in the passthrough commit. I can't follow much of the logic, but at least I didn't see anything that looks wrong.

kevmitch commented:
ok, removed the passthrough check. this was really just a sanity check as for passthrough, the WAVEFORMATEX is always set by set_waveformat, never by WASAPI.

I also changed the sample formats array to be hopefully more concise.

I'll squash these last two into the first on actual merge.

kevmitch commented:
squashed and merged minus 0e6759227f7132a92702418700548f4d339e8f76 as the anon struct was considered too weird.
